load.go

```diff
@@ -237,6 +239,61 @@ func LoadSpec(dt []byte) (*Spec, error) {
 	return &spec, nil
 }
 
+// rawYAML is similar to json.RawMesage
+// We use this to store the raw yaml data for extension fields
+type rawYAML []byte
+
+func (y rawYAML) MarshalYAML() ([]byte, error) {
+	return y, nil
+}
+
+func (y *rawYAML) UnmarshalYAML(dt []byte) error {
+	*y = dt
+	return nil
+}
+
+func (f *extensionFields) UnmarshalYAML(dt []byte) error {
+	// We need to store the raw yaml data for each extension key.
+	// Parse the yaml, grab all the extension keys and store the raw values in f.
+
+	parsed, err := parser.ParseBytes(dt, parser.ParseComments)
+	if err != nil {
+		return errors.Wrapf(err, "error parsing yaml: \n%s", string(dt))
+	}
+
+	if len(parsed.Docs) != 1 {
+		return errors.New("expected exactly one yaml document")
+	}
+
+	doc := parsed.Docs[0]
+	if doc.Body == nil {
+		return nil
+	}
+
+	body, ok := doc.Body.(*ast.MappingNode)
+	if !ok {
+		return errors.Errorf("expected a mapping node, got %T", body)
+	}
+
+	var ext extensionFields
+	for _, v := range body.Values {
+		key := v.Key.String()
+		if !strings.HasPrefix(key, "x-") && !strings.HasPrefix(key, "X-") {
+			return errors.Errorf("extension mapping key %q must not start with x-", key)
+		}
+
+		if ext == nil {
+			ext = make(extensionFields)
+		}
+		ext[key] = rawYAML(v.Value.String())
+	}
+
+	if ext != nil {
+		*f = ext
+	}
+	return nil
```

Comment:
It seems we eventually panic when the extensions are nil later on on line 350 when we unmarshal the extension nodes:

```go
s.extensions = ext
 if len(ext) == 0 {
 	panic("ext should not be empty")
 }
 ```
 
Why do we do that in the Spec.UnmarshalYAML function and not here for the individual extension fields UnmarshalYAML function? Is this because some fields might be nil and that's ok for later on when we unmarshal all of the extension nodes and just need to know if there are no extensions at all?

Reply:
In this function its not an error to not have anything to unmarshal.
In the spec, by the time we are doing that panic we have already collected a list of `x-` fields and so `ext` must not be empty.